src/receiver.c

```diff
@@ -0,0 +1,668 @@
+/**
+ * @file receiver.c  Generic stream receiver
+ *
+ * Copyright (C) 2023 Alfred E. Heggestad, Christian Spielberger
+ */
+#include <string.h>
+#include <time.h>
+#include <re.h>
+#include <baresip.h>
+#include "core.h"
+
+/** Magic number */
+#define MAGIC 0x00511eb3
+#include "magic.h"
+
+/* Receive */
+struct receiver {
+#ifndef RELEASE
+	uint32_t magic;          /**< Magic number for debugging            */
+#endif
+	/* Data protected by mtx */
+	char *name;                    /**< Media name                       */
+	struct metric *metric;         /**< Metrics for receiving            */
+	struct jbuf *jbuf;             /**< Jitter Buffer for incoming RTP   */
+	bool enabled;                  /**< True if enabled                  */
+	uint64_t ts_last;              /**< Timestamp of last recv RTP pkt   */
+	uint32_t ssrc;                 /**< Incoming synchronization source  */
+	bool ssrc_set;                 /**< Incoming SSRC is set             */
+	uint32_t pseq;                 /**< Sequence number for incoming RTP */
+	bool pseq_set;                 /**< True if sequence number is set   */
+	bool rtp_estab;                /**< True if RTP stream established   */
+	bool run;                      /**< True if RX thread is running     */
+	mtx_t *mtx;                    /**< Mutex protects above fields      */
+
+	/* Unprotected data */
+	struct stream *strm;           /**< Stream                           */
+	struct rtp_sock *rtp;          /**< RTP Socket                       */
+	stream_pt_h *pth;              /**< Stream payload type handler      */
+	stream_rtp_h *rtph;            /**< Stream RTP handler               */
+	stream_rtpestab_h *rtpestabh;  /**< RTP established handler          */
+	void *arg;                     /**< Stream argument                  */
+	void *sessarg;                 /**< Session argument                 */
+	thrd_t thr;                    /**< RX thread                        */
+	struct tmr tmr;                /**< Timer for stopping RX thread     */
+	int pt;                        /**< Previous payload type            */
+};
+
+
+enum work_type {
+	WORK_RTCP,
+	WORK_RTPESTAB,
+	WORK_PTCHANGED,
+};
+
+
+struct work {
+	enum work_type type;
+	struct receiver *rx;
+	union {
+		struct rtcp_msg *rtcp;
+		struct {
+			uint8_t pt;
+			struct mbuf *mb;
+		} pt;
+	} u;
+};
+
+
+static void async_work_main(int err, void *arg);
+static void work_destructor(void *arg);
+
+
+/*
+ * functions that run in RX thread (if "rxmode thread" is configured)
+ */
+
+
+static void pass_rtcp_work(struct receiver *rx, const struct rtcp_msg *msg)
+{
+	struct work *w;
+	int err;
+
+	if (!rx->run) {
+		stream_process_rtcp(rx->strm, msg);
+		return;
+	}
+
+	w = mem_zalloc(sizeof(*w), work_destructor);
+	w->type    = WORK_RTCP;
+	w->rx      = rx;
+	mbuf_set_pos(msg->mb, 0);
+	err = rtcp_decode(&w->u.rtcp, msg->mb);
+	mbuf_skip_to_end(msg->mb);
```

Comment:
@sreimers Should we implement a deep copy function for `rtcp_msg` ?

Reply:
This solution isn't correct anyway. In `rtp.c` there is the `rtcp_recv_handler()`
```c
...
	while (0 == rtcp_decode(&msg, mb)) {

		/* handle internally first */
		rtcp_handler(rs->rtcp, msg);

		/* then relay to application */
		if (rs->rtcph)
			rs->rtcph(src, msg, rs->arg);

		mem_deref(msg);
	}
```
So it is possible that there are multiple RCTP msg in the `mbuf`. A deep copy function would contain the long `switch` statement similar to `rtcp_decode()`.

Reply:
Is there a write within rtcp handlers or aren't they use rtcp_msg read only and we can simply make it const and ensure lifetime of rtcp_msg is long enough?

Reply:
Deep copy for `rtcp_msg`: https://github.com/baresip/re/pull/698